File: J2EE_Frontend/src/services/cropRecommendationService.js
/**
 * Crop Recommendation Service
 * Service để gọi API gợi ý cây trồng từ ML model
 */

import { API_BASE_URL as CONFIG_API_BASE_URL, API_ENDPOINTS } from '../config/api.config';

// Use the backend API base URL directly
const API_BASE_URL = CONFIG_API_BASE_URL;

const cropRecommendationService = {
  /**
   * Gợi ý cây trồng dựa trên dữ liệu đất và môi trường
   * 
   * @param {Object} data - Dữ liệu đầu vào
   * @param {number} data.N - Nitrogen (Đạm) - ppm
   * @param {number} data.P - Phosphorus (Lân) - ppm
   * @param {number} data.K - Potassium (Kali) - ppm
   * @param {number} data.temperature - Nhiệt độ (°C)
   * @param {number} data.humidity - Độ ẩm (%)
   * @param {number} data.ph - Độ pH
   * @param {number} data.rainfall - Lượng mưa (mm)
   * @returns {Promise<Object>} Kết quả gợi ý
   */
  async recommendCrop(data) {
    try {
      // Call backend endpoint /api/crop/recommend
      const response = await fetch(`${API_BASE_URL}/api/crop/recommend`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${localStorage.getItem('token') || ''}`
        },
        body: JSON.stringify(data)
      });

      if (!response.ok) {
        const errorText = await response.text();
        let errorMessage = `HTTP error! status: ${response.status}`;
        try {
          const errorJson = JSON.parse(errorText);
          errorMessage = errorJson.error || errorJson.message || errorMessage;
        } catch (e) {
          errorMessage = errorText || errorMessage;
        }
        throw new Error(errorMessage);
      }

      const result = await response.json();
      
      // Log raw response để debug
      console.log('🔍 Raw crop recommendation response:', result);
      console.log('🔍 Response keys:', Object.keys(result));
      
      // Đảm bảo response có format đúng
      // Backend trả về: { success: true, recommended_crop: "...", crop_name_en: "...", confidence: 0.8, input_data: {...} }
      if (!result.success && !result.recommended_crop) {
        console.warn('⚠️ Response không có success hoặc recommended_crop');
        return {
          success: false,
          error: result.error || 'Không thể nhận được gợi ý từ server'
        };
      }
      
      // Đảm bảo có recommended_crop (thử nhiều field names)
      if (result.success && !result.recommended_crop) {
        console.warn('⚠️ Response success nhưng không có recommended_crop, tìm fallback...');
        result.recommended_crop = result.crop || 
                                 result.recommendedCrop || 
                                 result.crop_name || 
                                 result.cropName ||
                                 'Cây trồng được gợi ý';
        console.log('✅ Fallback crop name:', result.recommended_crop);
      }
      
      // Đảm bảo có input_data nếu không có
      if (result.success && !result.input_data) {
        result.input_data = {
          temperature: result.temperature || data.temperature || '',
          humidity: result.humidity || data.humidity || '',
          soil_moisture: result.soil_moisture || data.soil_moisture || ''
        };
      }
      
      console.log('✅ Final processed result:', result);
      return result;
    } catch (error) {
      console.error('Error recommending crop:', error);
      return {
        success: false,
        error: error.message || 'Không thể kết nối đến server. Vui lòng thử lại sau.'
      };
    }
  },

  /**
   * Gợi ý cây trồng cho nhiều mẫu
   * 
   * @param {Array} samples - Mảng các mẫu dữ liệu
   * @returns {Promise<Object>} Kết quả gợi ý batch
   */
  async recommendCropBatch(samples) {
    try {
      const response = await fetch(`${API_BASE_URL}/api/crop/recommend-batch`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ samples })
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      const result = await response.json();
      return result;
    } catch (error) {
      console.error('Error recommending crops (batch):', error);
      throw error;
    }
  },

  /**
   * Lấy danh sách các loại cây trồng
   * 
   * @returns {Promise<Object>} Danh sách cây trồng
   */
  async getCropList() {
    try {
      const response = await fetch(`${API_BASE_URL}/api/crop/crops`);

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      const result = await response.json();
      return result;
    } catch (error) {
      console.error('Error getting crop list:', error);
      throw error;
    }
  },

  /**
   * Kiểm tra trạng thái ML service
   * 
   * @returns {Promise<Object>} Trạng thái service
   */
  async checkHealth() {
    try {
      const response = await fetch(`${API_BASE_URL}/api/crop/health`);

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      const result = await response.json();
      return result;
    } catch (error) {
      console.error('Error checking ML service health:', error);
      return {
        status: 'unhealthy',
        model_loaded: false,
        error: error.message
      };
    }
  }
};

export default cropRecommendationService;

File: RecommentCrop/crop_recommendation_service.py
"""
Crop Recommendation Service
API service để gợi ý cây trồng dựa trên model RandomForest
"""

import pickle
import numpy as np
from flask import Flask, request, jsonify
from flask_cors import CORS
import logging
import os

# Cấu hình logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

app = Flask(__name__)

# Configure CORS with environment variable support
origins_env = os.environ.get("FRONTEND_ORIGINS", "")
if origins_env:
    origins = [origin.strip() for origin in origins_env.split(",") if origin.strip()]
else:
    # Default origins for local development
    origins = [
        "http://localhost:3000",
        "http://localhost:8000",
        "http://localhost:8080",
        "http://localhost:9002"
    ]

CORS(app, origins=origins, supports_credentials=True, 
     allow_headers=["Content-Type", "Authorization"],
     methods=["GET", "POST", "PUT", "DELETE", "OPTIONS"])

# Load model khi khởi động service
MODEL_PATH = 'RandomForest_RecomentTree.pkl'
model = None

# Mapping crop names (English to Vietnamese)
# Model trả về tên tiếng Anh, cần chuyển sang tiếng Việt
CROP_NAMES = {
    'apple': 'Táo',
    'banana': 'Chuối',
    'blackgram': 'Đậu đen',
    'chickpea': 'Đậu gà',
    'coconut': 'Dừa',
    'coffee': 'Cà phê',
    'cotton': 'Bông',
    'grapes': 'Nho',
    'jute': 'Đay',
    'kidneybeans': 'Đậu thận',
    'lentil': 'Đậu lăng',
    'maize': 'Ngô',
    'mango': 'Xoài',
    'mothbeans': 'Đậu ngài',
    'mungbean': 'Đậu xanh',
    'muskmelon': 'Dưa lưới',
    'orange': 'Cam',
    'papaya': 'Đu đủ',
    'pigeonpeas': 'Đậu chim',
    'pomegranate': 'Lựu',
    'rice': 'Lúa',
    'watermelon': 'Dưa hấu'
}

def load_model():
    """Load RandomForest model từ file pkl"""
    global model
    try:
        # Kiểm tra file tồn tại
        if not os.path.exists(MODEL_PATH):
            logger.error(f"Model file không tồn tại: {MODEL_PATH}")
            logger.error(f"Current working directory: {os.getcwd()}")
            logger.error(f"Files in current directory: {os.listdir('.')}")
            return False
        
        # Kiểm tra file size
        file_size = os.path.getsize(MODEL_PATH)
        logger.info(f"Model file size: {file_size} bytes")
        
        if file_size < 100:
            logger.error(f"Model file quá nhỏ ({file_size} bytes), có thể bị corrupt")
            return False
        
        # Kiểm tra file có phải là binary không (đọc vài byte đầu)
        try:
            with open(MODEL_PATH, 'rb') as f:
                first_bytes = f.read(10)
                logger.info(f"First 10 bytes (hex): {first_bytes.hex()}")
                # Pickle files thường bắt đầu với các byte đặc biệt
                if first_bytes.startswith(b'version') or first_bytes.startswith(b'v'):
                    logger.error("File có vẻ là text file, không phải pickle binary")
                    return False
        except Exception as e:
            logger.error(f"Không thể đọc file: {str(e)}")
            return False
        
        # Try using joblib first (better for sklearn models)
        try:
            import joblib
            logger.info("🔄 Thử load model bằng joblib...")
            model = joblib.load(MODEL_PATH)
            logger.info("✓ Model loaded using joblib")
        except Exception as joblib_error:
            logger.warning(f"Joblib load failed: {str(joblib_error)}")
            logger.info("🔄 Thử load model bằng pickle...")
            
            # Fallback to pickle
            import warnings
            warnings.filterwarnings('ignore', category=UserWarning)
            
            try:
                with open(MODEL_PATH, 'rb') as f:
                    model = pickle.load(f)
                logger.info("✓ Model loaded using pickle")
            except Exception as pickle_error:
                logger.error(f"Pickle load failed: {str(pickle_error)}")
                import traceback
                traceback.print_exc()
                return False
        
        logger.info("✓ Model đã được load thành công!")
        logger.info(f"Model type: {type(model)}")
        
        # Kiểm tra model có method predict không
        if hasattr(model, 'predict'):
            logger.info("✓ Model có method predict()")
        else:
            logger.error("Model không có method predict()")
            return False
        
        return True
    except Exception as e:
        logger.error(f"Lỗi khi load model: {str(e)}")
        import traceback
        traceback.print_exc()
        logger.error("Gợi ý: Model có thể được train với sklearn version khác. Cần train lại model với sklearn hiện tại.")
        return False

@app.route('/health', methods=['GET'])
def health_check():
    """Health check endpoint"""
    return jsonify({
        'status': 'healthy',
        'model_loaded': model is not None
    }), 200

@app.route('/api/recommend-crop', methods=['POST'])
def recommend_crop():
    """
    API endpoint để gợi ý cây trồng
    
    Request body:
    {
        "temperature": 25.5,    # Nhiệt độ (°C)
        "humidity": 80.0,       # Độ ẩm không khí (%)
        "soil_moisture": 45.0   # Độ ẩm đất (%)
    }
    
    Response:
    {
        "success": true,
        "recommended_crop": "Lúa",
        "crop_name_en": "rice",
        "confidence": 0.95,
        "input_data": {...}
    }
    """
    try:
        if model is None:
            return jsonify({
                'success': False,
                'error': 'Model chưa được load'
            }), 500
        
        data = request.get_json()
        
        # Validate input data - 3 FEATURES: Temperature, Humidity, Soil_Moisture
        required_fields = ['temperature', 'humidity', 'soil_moisture']
        for field in required_fields:
            if field not in data:
                return jsonify({
                    'success': False,
                    'error': f'Thiếu trường bắt buộc: {field}'
                }), 400
        
        # Chuẩn bị dữ liệu input cho model
        # Model nhận 3 features: Temperature, Humidity, Soil_Moisture
        input_features = np.array([[
            float(data['temperature']),
            float(data['humidity']),
            float(data['soil_moisture'])
        ]])
        
        # Dự đoán - Model trả về tên cây trồng bằng tiếng Anh
        prediction = model.predict(input_features)[0]
        
        # Nếu model hỗ trợ predict_proba, lấy confidence
        confidence = None
        if hasattr(model, 'predict_proba'):
            probabilities = model.predict_proba(input_features)[0]
            confidence = float(max(probabilities))
        
        # Chuyển đổi tên tiếng Anh sang tiếng Việt
        crop_name_en = str(prediction).lower()
        crop_name_vi = CROP_NAMES.get(crop_name_en, prediction)
        
        logger.info(f"Dự đoán: {crop_name_vi} ({crop_name_en}), Confidence: {confidence}")
        
        return jsonify({
            'success': True,
            'recommended_crop': crop_name_vi,
            'crop_name_en': crop_name_en,
            'confidence': confidence,
            'input_data': data
        }), 200
        
    except ValueError as e:
        return jsonify({
            'success': False,
            'error': f'Dữ liệu đầu vào không hợp lệ: {str(e)}'
        }), 400
    except Exception as e:
        logger.error(f"Lỗi khi dự đoán: {str(e)}")
        return jsonify({
            'success': False,
            'error': f'Lỗi server: {str(e)}'
        }), 500

@app.route('/api/recommend-crop/batch', methods=['POST'])
def recommend_crop_batch():
    """
    API endpoint để gợi ý cây trồng cho nhiều mẫu
    
    Request body:
    {
        "samples": [
            {"temperature": 25.5, "humidity": 80.0, "soil_moisture": 45.0},
            {"temperature": 28.0, "humidity": 75.0, "soil_moisture": 50.0},
            {...}
        ]
    }
    """
    try:
        if model is None:
            return jsonify({
                'success': False,
                'error': 'Model chưa được load'
            }), 500
        
        data = request.get_json()
        samples = data.get('samples', [])
        
        if not samples:
            return jsonify({
                'success': False,
                'error': 'Không có dữ liệu samples'
            }), 400
        
        results = []
        for idx, sample in enumerate(samples):
            try:
                # Model nhận 3 features: Temperature, Humidity, Soil_Moisture
                input_features = np.array([[
                    float(sample['temperature']),
                    float(sample['humidity']),
                    float(sample['soil_moisture'])
                ]])
                
                prediction = model.predict(input_features)[0]
                
                confidence = None
                if hasattr(model, 'predict_proba'):
                    probabilities = model.predict_proba(input_features)[0]
                    confidence = float(max(probabilities))
                
                # Chuyển đổi tên tiếng Anh sang tiếng Việt
                crop_name_en = str(prediction).lower()
                crop_name_vi = CROP_NAMES.get(crop_name_en, prediction)
                
                results.append({
                    'index': idx,
                    'recommended_crop': crop_name_vi,
                    'crop_name_en': crop_name_en,
                    'confidence': confidence,
                    'success': True
                })
            except Exception as e:
                results.append({
                    'index': idx,
                    'success': False,
                    'error': str(e)
                })
        
        return jsonify({
            'success': True,
            'results': results
        }), 200
        
    except Exception as e:
        logger.error(f"Lỗi khi dự đoán batch: {str(e)}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@app.route('/api/crops', methods=['GET'])
def get_crop_list():
    """Lấy danh sách tất cả các loại cây trồng"""
    crops = [{'name_en': name_en, 'name_vi': name_vi} for name_en, name_vi in CROP_NAMES.items()]
    return jsonify({
        'success': True,
        'total': len(crops),
        'crops': crops
    }), 200

if __name__ == '__main__':
    # Load model khi khởi động (nhưng vẫn chạy service dù không load được)
    load_model()
    
    logger.info("Đang khởi động Crop Recommendation Service...")
    logger.info("API sẽ chạy tại: http://localhost:5000")
    logger.info("\nEndpoints available:")
    logger.info("  - GET  /health                    - Health check")
    logger.info("  - POST /api/recommend-crop        - Gợi ý cây trồng (single)")
    logger.info("  - POST /api/recommend-crop/batch  - Gợi ý cây trồng (batch)")
    logger.info("  - GET  /api/crops                 - Danh sách cây trồng")
    
    if model is None:
        logger.warning("⚠️  WARNING: Model chưa được load. API /api/recommend-crop sẽ trả về lỗi.")
        logger.warning("⚠️  Service vẫn chạy để health check hoạt động.")
    
    app.run(host='0.0.0.0', port=5000, debug=False)


